### PR TITLE
feat: support memory copying instruction (EIP-5656)

### PR DESCRIPTION
**Pull request by Agusrodri**
This PR adds an adaptation to [EIP-5656](https://eips.ethereum.org/EIPS/eip-5656) which introduces the `MCOPY` opcode. 

By using this new instruction, it is possible to copy parts of the memory in a cheaper and easier way, everything done with one opcode.

### Note

All the Cancun [ethereum-tests](https://github.com/ethereum/tests/tree/428f218d7d6f4a52544e12684afbfe6e2882ffbf/GeneralStateTests/Cancun/stEIP5656-MCOPY) related to `MCOPY` pass with this implementation.

### CODE REVIEW

**Comment by RomarQ** [comment]
We could add a new flag `eip_5656_enabled` to the config for reverting when `MCOPY` is used in forks that should not support it.

Similar change for eip-1153 https://github.com/rust-ethereum/evm/pull/278/commits/b8bfe595de0c351c18a17546e6b9475d093fba99

**Comment by Agusrodri** [comment]
> We could add a new flag `eip_5656_enabled` to the config for reverting when `MCOPY` is used in forks that should not support it.
> 
> Similar change for eip-1153 [b8bfe59](https://github.com/rust-ethereum/evm/commit/b8bfe595de0c351c18a17546e6b9475d093fba99)

Done in 3f15f922d23df2f7453c626de5ccdc9c3094c270 :)

**Comment by koushiro** [comment]
@Agusrodri please resolve the conflicts

**Comment by RomarQ** [comment]
> @Agusrodri please resolve the conflicts

Done :+1:

File: interpreter/src/machine/memory.rs [hunk 1]
```diff
@@ -266,4 +275,49 @@ mod tests {
...
+		memory.set(3usize, &[1u8, 2u8, 3u8, 4u8], None).unwrap();
+		assert_eq!(memory.data(), &[0u8, 0u8, 0u8, 1u8, 2u8, 3u8, 4u8].to_vec());
+
+		// Copy 2 bytes into index 3.
+		// As the length is 2, we copy the bytes present on indexes 6 and 7,
+		// which are [4,0].
+		memory.copy(3usize, 6usize, 2usize);
+
+		// Now the new memory data results in [0, 0, 0, 4, 0, 3, 4, 0].
+		// An extra element is added due to rezising.
```
Comment (koushiro) [inline interpreter/src/machine/memory.rs:317]:
```suggestion
		// An extra element is added due to resizing.
```

**Review by koushiro — APPROVED** [review]

**Pull request merged**